credentials/apps/badges/base_api_client.py

```diff
@@ -0,0 +1,93 @@
+import logging
+from abc import ABC, abstractmethod
+from urllib.parse import urljoin
+
+import requests
+from requests.exceptions import HTTPError
+
+from .exceptions import BadgeProviderError
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseBadgeProviderClient(ABC):
+    """
+    Base class for interacting with a generic badge provider API.
+
+    This class provides common functionality such as performing requests
+    and error handling. Methods specific to badge providers must be implemented
+    by subclasses.
+    """
+
+    PROVIDER_NAME = None
+
+    def __init__(self):
+        self.base_api_url = None
```

Comment:
Perhaps it is worth making `base_api_url` a class attribute or property for more convenient overrides?

Reply:
This field depends on dynamic value in Credly implementation
https://github.com/raccoongang/credentials/pull/186/files#diff-0013cdcb6c06dc151ce666d7c31fead0fbfaddbe9e9d2a470ecd1288e006b1d5R46

Reply:
It can be a class property that dynamically calculates the url. Why not?
```suggestion
class BaseBadgeProviderClient(ABC):
        ...
        
        @property    
        def base_api_url(self):
            pass    
        
class CredlyAPIClient(BaseBadgeProviderClient):
        ...
                
        @property
        def base_api_url(self):
            return urljoin(get_credly_api_base_url(settings), f"organizations/{self.organization_id}/")
```
or 
```suggestion
class BaseBadgeProviderClient(ABC):
        ...
        
        @property    
        def base_api_url(self):
            return self._get_base_api_url()   
                         
        def _get_base_api_url(self):
             pass
        
class CredlyAPIClient(BaseBadgeProviderClient):
        ...
                
        def _get_base_api_url(self):
            return urljoin(get_credly_api_base_url(settings), f"organizations/{self.organization_id}/")   
```